src/amqp/amqppublishnotifier.cpp

```diff
@@ -0,0 +1,182 @@
+void AMQPAbstractPublishNotifier::SpawnProtonContainer()
+{
+    try {
+        proton::default_container(*handler_).run();
+    }
+    catch (const proton::error_condition &e) {
+        LogPrint("amqp", "amqp: container error: %s\n", e.what());
+    }
+    catch (const std::runtime_error &e) {
+        LogPrint("amqp", "amqp: runtime error: %s\n", e.what());
+    }
+    catch (const std::exception &e) {
+        LogPrint("amqp", "amqp: exception: %s\n", e.what());
+    }
+    catch (...) {
+        LogPrint("amqp", "amqp: unknown error\n");
+    }
+    handler_->terminate();
+}
+
+bool AMQPAbstractPublishNotifier::Initialize()
+{
+    std::multimap<std::string, AMQPAbstractPublishNotifier*>::iterator i = mapPublishNotifiers.find(address);
+
+    if (i==mapPublishNotifiers.end()) {
+        try {
+            handler_ = std::make_shared<amqp_sender>(address);
+            thread_ = std::make_shared<std::thread>(&AMQPAbstractPublishNotifier::SpawnProtonContainer, this);
+        }
+        catch (std::exception &e) {
+            LogPrint("amqp", "amqp: initialization error: %s\n", e.what());
+            return false;
+        }
+        mapPublishNotifiers.insert(std::make_pair(address, this));
+    } else {
+        // copy the shared ptrs to the message handler and the thread where the proton container is running
+        handler_ = i->second->handler_;
+        thread_ = i->second->thread_;
+        mapPublishNotifiers.insert(std::make_pair(address, this));
+    }
+
+    return true;
+}
+
+
+void AMQPAbstractPublishNotifier::Shutdown()
+{
+    LogPrint("amqp", "amqp: Shutdown notifier %s at %s\n", GetType(), GetAddress());
+
+    int count = mapPublishNotifiers.count(address);
+
+    // remove this notifier from the list of publishers using this address
+    typedef std::multimap<std::string, AMQPAbstractPublishNotifier*>::iterator iterator;
+    std::pair<iterator, iterator> iterpair = mapPublishNotifiers.equal_range(address);
+
+    for (iterator it = iterpair.first; it != iterpair.second; ++it) {
+        if (it->second == this) {
+            mapPublishNotifiers.erase(it);
+            break;
+        }
+    }
+
+    // terminate the connection if this is the last publisher using this address
+    if (count == 1) {
+        handler_->terminate();
+        if (thread_.get() != nullptr) {
+            if (thread_->joinable()) {
+                thread_->join();
+            }
+        }
+    }
+}
+
+
+bool AMQPAbstractPublishNotifier::SendMessage(const char *command, const void* data, size_t size)
+{
+    if (handler_->isTerminated()) {
+        return false;
+    }
+
+    try { 
+        proton::binary content;
+        const char *p = (const char *)data;
+        content.assign(p, p + size);
+
+        proton::message message(content);
+        message.subject(std::string(command));
+        proton::message::property_map & props = message.properties();
+        props.put("x-opt-sequence-number", sequence_);
```

Comment:
Yes, and there are other potential useful fields, but I'd like to have that in another PR as an improvement.  This PR is bringing AMQP notifications upto speed with the ZMQ notifications.